### PR TITLE
[WIP][SYSTEMDS-2782] Built-in mdedup

### PR DESCRIPTION
**Pull request by OlgaOvcharenko**
This PR adds a new built-in mdedup for detecting duplicates in frames using matching dependencies (like Street 0.95, City 0.90 -> ZIP 1.0). 
@Shafaq-Siddiqi For simplicity, I used Jaccard similarity, but found out that Levenshtein or Jaro distance could also be used, should I also add them? To compute Jaccard similarity between rows (strings) of a vector (nx1) the map with 2 args was added dist = map(Xi, "(x, y) -> UtilFunctions.jaccardSim(x, y)").

I also modified discoverFD built-in by setting diag to 1.

### CODE REVIEW

**Comment by Shafaq-Siddiqi** [comment]
> This PR adds a new built-in mdedup for detecting duplicates in frames using matching dependencies (like Street 0.95, City 0.90 -> ZIP 1.0).
> @Shafaq-Siddiqi For simplicity, used Jaccard similarity, but if found out that Levenshtein or Jaro distance could also be used, should I also add them? To compute Jaccard similarity between rows (strings) of a vector (nx1) the map with 2 args was added dist = map(Xi, "(x, y) -> UtilFunctions.jaccardSim(x, y)").
> 
> I also modified discoverFD built-in by setting diag to 1.

Thanks @OlgaOvcharenko, for now it is fine to have Jaccard similarity only we will come to other methods when we will extend the overall implementation.

File: scripts/builtin/mdedup.dml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+  MD = matrix(0, n, 1)
+
+  LHS_MD = getMDAdjacency(X, LHSfeatures, LHSthreshold)
+
+  if (sum(LHS_MD) > 0) {
+    RHS_MD = getMDAdjacency(X, RHSfeatures, RHSthreshold)
+  }
+
+  MD = detectDuplicates(LHS_MD, LHS_MD)
```
Comment (Shafaq-Siddiqi) [inline scripts/builtin/mdedup.dml:71]: Probably a typo.

File: scripts/builtin/mdedup.dml [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    threshold = as.scalar(thresholds[1, i])
+#print(toString(jaccardDist))
+    if(i == 1) {
+      adjacency = jaccardDist >= threshold
+    } else {
+      adjacency = adjacency & (jaccardDist >= threshold)
+    }
+#print(toString(adjacency))
+    # break if one of MDs is false
+    if (sum(adjacency) == 0)
```
Comment (Shafaq-Siddiqi) [inline scripts/builtin/mdedup.dml:101]: explicitly setting the value of for loop variable will not break the for loop here so if you want to terminate a loop early use a while loop please.

File: src/main/java/org/apache/sysds/runtime/matrix/data/FrameBlock.java [hunk 3]
```diff
@@ -2147,8 +2166,45 @@ public static FrameMapFunction getCompiledFunction(String lambdaExpr) {
 		}
 	}
 
+	public static FrameMapDistFunction getCompiledFunction2Args(String lambdaExpr) {
+		// split lambda expression
+		String[] parts = lambdaExpr.split("->");
+		parts[0] = parts[0].replace("(", "");
```
Comment (Shafaq-Siddiqi) [inline src/main/java/org/apache/sysds/runtime/matrix/data/FrameBlock.java:2172]: these tow lines could be replaced with the following 
parts[0] = parts[0].replaceAll("[()]", "");

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinMDTest.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(value = Parameterized.class)
+public class BuiltinMDTest extends AutomatedTestBase {
+	private final static String TEST_NAME = "matching_dependency";
+	private final static String TEST_DIR = "functions/builtin/";
+	private static final String TEST_CLASS_DIR = TEST_DIR + BuiltinMDTest.class.getSimpleName() + "/";
+	private final static Types.ValueType[] schemaStrings = {
+		Types.ValueType.STRING, Types.ValueType.STRING, Types.ValueType.STRING,
```
Comment (Shafaq-Siddiqi) [inline src/test/java/org/apache/sysds/test/functions/builtin/BuiltinMDTest.java:46]: could be replaced with 
private final static Types.ValueType[] schemaStrings = UtilFunctions.nCopies(cols, Types.ValueType.STRING);

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinMDTest.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+	@Test
+	public void testMD1() {
+		double[][] D =  {
+			{7567, 231, 1231, 1232, 122, 321},
+			{5321, 23123, 122, 123, 1232, 11},
+			{7267, 3, 223, 432, 1132, 0},
+			{7267, 3, 223, 432, 1132, 500},
+			{7254, 3, 223, 432, 1132, 0},
+		};
```
Comment (Shafaq-Siddiqi) [inline src/test/java/org/apache/sysds/test/functions/builtin/BuiltinMDTest.java:95]: could you please also create a test case for Spark?

**Comment by OlgaOvcharenko** [comment]
Thanks @Shafaq-Siddiqi , I fixed this issues

**Comment by Shafaq-Siddiqi** [comment]
Hi @OlgaOvcharenko ,
Thanks for fixing the issues but the Spark tests are producing invalid results (i.e., all zeros). Could you please have a look into this?

File: scripts/builtin/mdedup.dml [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  adjacency = matrix(0, n, n)
+
+  i = 1
+  while (i <= ncol(features)) {
+    # slice col
+    pos = as.scalar(features[1, i])
+    Xi = X[, pos]
+
+    # distances between words in each row of col
+    dist = map(Xi, "(x, y) -> UtilFunctions.jaccardSim(x, y)")
```
Comment (OlgaOvcharenko) [inline scripts/builtin/mdedup.dml:91]: dist matrix has nx1 dimensions, but dims were set to nxn in BinaryFrameScalarSPInstruction.java and BuiltinFunctionExpression.java. This is the reason why components fail and thus mdeup in SPARK.

**Comment by Shafaq-Siddiqi** [comment]
LGTM. Thank you @OlgaOvcharenko for the contribution. During the merge, I fixed some formatting issues, simplified the map() code for accepting two parameters. I have temporarily fixed the Spark tests by specifying the matrix indices inside the component function call. It seems to be a bug in size propagation I will create a JIRA report for this.

**Pull request closed without merging**